MDLSITE-3782 precheck javascript with ESLint when present

danpoltawski opened the pull request:
ESLint superseeds jshint, when MDL-52127 lands we'll have much better JS coding style checking. This adds patch support to the prechecker whilst providing BC with old branches with jshint. 

After we've given this a bit of time to settle I actually would propose we jshint from the prechecker completely, and make ESlint work on older branches instead (we'd still get lint jshint reported by grunt  for amd/yui files). See also MDL-54889.

danpoltawski commented:
You can test this now with my branch:
git pull git://github.com/danpoltawski/moodle.git eslint-prechecker

danpoltawski commented on remote_branch_checker/remote_branch_checker.sh:
```diff
@@ -405,9 +407,13 @@ ${phpcmd} ${mydir}/../../moodlecheck/cli/moodlecheck.php \
-${jshintcmd} --config $WORKSPACE/.jshintrc --exclude-path $WORKSPACE/.jshintignore \
-    --reporter=checkstyle ${WORKSPACE} > "${WORKSPACE}/work/jshint.xml"
+if [ -f $WORKSPACE/.eslintrc ]; then
+    ./node_modules/.bin/eslint -f checkstyle $WORKSPACE > "${WORKSPACE}/work/eslint.xml"
```
This raw use of node_modules might be controversial, thoughts welcome. (I think we should use the grunt installed version because we then don't need to worry about version dependencies.

stronk7 commented:
As commented in the chat... what if we use ```$(${npmcmd} bin))``` to know where to find the executables?

danpoltawski commented:
> As commented in the chat... what if we use $(${npmcmd} bin)) to know where to find the executables?

I have updated the patch to do that (and also error more gracefully if the npm file isn't found)

danpoltawski commented:
Note that I updated the eslint-prechecker branch on top of latest in MDL-52127

stronk7 commented:
Ping, maybe you should move this on top of #85, for your consideration, ciao :-)

danpoltawski commented:
I have rebased on top of #85 - I changed the approach slightly different to what it was before - instead of the if/else I do the eslint stuff before the node_modules directory is deleted, and then check for the presenece of eslint.xml

stronk7 commented:
Installed in the laptop, running a couple of tests (with and without .eslintrc) ... before merging this into production...

stronk7 commented:
Worked ok against your branch adding .eslintrc, grunt installed it ok and warnings were detected:

http://ci.stronk7.com/view/prehecker/job/Precheck%20remote%20branch/1710/

stronk7 commented:
And also ok against current master, fall-backing to jshint:

http://ci.stronk7.com/view/prehecker/job/Precheck%20remote%20branch/1711/

stronk7 commented:
So, merged! Let's break the unheld issues... thanks!
